File: research/explorations/bench-realistic/strings.lua
local function bench(n)
    local s = 0
    for i = 0, n - 1 do
        local t = "item-" .. tostring(i)
        -- split on "-"
        local parts = {}
        for part in t:gmatch("[^-]+") do
            parts[#parts + 1] = part
        end
        local j = table.concat(parts, "_")
        s = s + #j
    end
    return s
end

local N = tonumber(arg[1]) or 1000
-- warmup
for i = 1, 100 do bench(N) end
local iters = 10000
local clock = os.clock
local start = clock()
local r
for i = 1, iters do r = bench(N) end
local elapsed = (clock() - start) * 1e9
io.write(string.format("result:     %d\n", r))
io.write(string.format("iterations: %d\n", iters))
io.write(string.format("total:      %.2fms\n", elapsed / 1e6))
io.write(string.format("per call:   %.0fns\n", elapsed / iters))

File: research/explorations/bench-realistic/strings.rs
use std::env;
use std::time::Instant;

fn bench(n: i64) -> i64 {
    let mut s: i64 = 0;
    for i in 0..n {
        let t = format!("item-{}", i);
        let parts: Vec<&str> = t.split('-').collect();
        let j = parts.join("_");
        s += j.len() as i64;
    }
    s
}

fn main() {
    let n: i64 = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(1000);
    for _ in 0..1000 { let _ = bench(n); }
    let iters: u128 = 10000;
    let start = Instant::now();
    let mut r = 0i64;
    for _ in 0..iters { r = bench(n); }
    let elapsed = start.elapsed().as_nanos();
    println!("result:     {}", r);
    println!("iterations: {}", iters);
    println!("total:      {:.2}ms", elapsed as f64 / 1e6);
    println!("per call:   {}ns", elapsed / iters);
}
